db/multi_cf_iterator_impl.cc

```diff
@@ -0,0 +1,206 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include <cassert>
+
+#include "rocksdb/multi_cf_iterator.h"
+#include "util/heap.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class MultiCfIteratorImpl : public MultiCfIterator {
+ public:
+  MultiCfIteratorImpl() {}
+  MultiCfIteratorImpl(const Comparator* comparator,
+                      const std::vector<ColumnFamilyHandle*>& column_families,
+                      const std::vector<Iterator*>& child_iterators)
+      : comparator_(comparator),
+        min_heap_(MultiCfMinHeapItemComparator(comparator_)) {
+    assert(column_families.size() > 0);
+    assert(column_families.size() == child_iterators.size());
+
+    cfhs_.reserve(column_families.size());
+    iterators_.reserve(column_families.size());
+    for (size_t i = 0; i < column_families.size(); ++i) {
+      cfhs_.push_back(column_families[i]);
+      iterators_.push_back(child_iterators[i]);
+    }
+  }
+  ~MultiCfIteratorImpl() {
+    for (auto iter : iterators_) {
+      delete iter;
+    }
+    status_.PermitUncheckedError();
+  }
+
+  // No copy allowed
+  MultiCfIteratorImpl(const MultiCfIteratorImpl&) = delete;
+  MultiCfIteratorImpl& operator=(const MultiCfIteratorImpl&) = delete;
+
+  Slice key() const override {
+    assert(Valid());
+    return min_heap_.top().iterator->key();
+  }
+  bool Valid() const override { return !min_heap_.empty() && status_.ok(); }
+  Status status() const override { return status_; }
+  void considerStatus(Status s) {
+    if (!s.ok() && status_.ok()) {
+      status_ = s;
+    }
+  }
+  void Reset() {
+    min_heap_.clear();
+    status_ = Status::OK();
+  }
+  void SeekToFirst() override {
+    Reset();
+    int i = 0;
+    for (auto& iter : iterators_) {
+      auto& cfh = cfhs_[i];
+      iter->SeekToFirst();
+      if (iter->Valid()) {
+        assert(iter->status().ok());
+        min_heap_.push(MultiCfIteratorInfo{iter, cfh, i});
+      } else {
+        considerStatus(iter->status());
+      }
+      ++i;
+    }
+  }
+
+  void Next() override {
+    assert(Valid());
+
+    auto* current = min_heap_.top().iterator;
+    std::string current_key_copy =
```

Comment:
Here, we first store the key from the iterator at the top of the heap. Then the top iterator moves to the next element and heap gets re-heapified. If the same key appears at the top of the heap, it repeats. Basically, this Next() finds the next key in order across all the iterators in the heap by skipping the same key that we've seen.

I'd love to see if there's a better way to achieve this (e.g. without copying the first key)